src/dbnode/encoding/arrow/tile/series_block_iterator.go

```diff
@@ -0,0 +1,170 @@
+	Next() bool
+	Close() error
+	Current() []SeriesFrameIterator
+	Err() error
+}
+
+type seriesBlockIter struct {
+	reader fs.DataFileSetReader
+
+	err           error
+	exhausted     bool
+	hasLastValues bool
+	concurrency   int
+	freedAfter    int
+
+	step  xtime.UnixNano
+	start xtime.UnixNano
+
+	encodingOpts encoding.Options
+	recorders    []*datapointRecorder
+	iters        []SeriesFrameIterator
+	byteReaders  []*bytes.Reader
+	baseIters    []encoding.ReaderIterator
+	bytesRefHeld []bool
+	dataBytes    []checked.Bytes
+	tagIters     []ident.TagIterator
+}
+
+// NewSeriesBlockIterator creates a new SeriesBlockIterator.
+func NewSeriesBlockIterator(
+	reader fs.DataFileSetReader,
+	step xtime.UnixNano,
+	start xtime.UnixNano,
+	concurrency int,
+	encodingOpts encoding.Options,
+) (SeriesBlockIterator, error) {
+	if concurrency < 1 {
+		return nil, fmt.Errorf("concurrency must be greater than 1, is: %d", concurrency)
+	}
+
+	var (
+		recorders   = make([]*datapointRecorder, 0, concurrency)
+		iters       = make([]SeriesFrameIterator, 0, concurrency)
+		byteReaders = make([]*bytes.Reader, 0, concurrency)
+		baseIters   = make([]encoding.ReaderIterator, 0, concurrency)
+	)
+
+	for i := 0; i < concurrency; i++ {
+		recorder := newDatapointRecorder(memory.NewGoAllocator())
+		recorders = append(recorders, recorder)
+		iters = append(iters, newSeriesFrameIterator(recorder))
+		byteReaders = append(byteReaders, bytes.NewReader(nil))
+		baseIters = append(baseIters, m3tsz.NewReaderIterator(nil, true, encodingOpts))
+	}
+
+	return &seriesBlockIter{
+		reader: reader,
+
+		concurrency: concurrency,
+		freedAfter:  concurrency,
+		start:       start,
+		step:        step,
+
+		encodingOpts: encodingOpts,
+		recorders:    recorders,
+		iters:        iters,
+		byteReaders:  byteReaders,
+		baseIters:    baseIters,
+		bytesRefHeld: make([]bool, concurrency),
+		dataBytes:    make([]checked.Bytes, concurrency),
+		tagIters:     make([]ident.TagIterator, concurrency),
+	}, nil
+}
+
+func (b *seriesBlockIter) Next() bool {
+	if b.exhausted || b.err != nil {
+		return false
+	}
+
+	for i, held := range b.bytesRefHeld {
+		if held && b.dataBytes[i] != nil {
+			b.bytesRefHeld[i] = false
+			b.dataBytes[i].DecRef()
+		}
+	}
+
+	var err error
+	for i := 0; i < b.concurrency; i++ {
+		_, b.tagIters[i], b.dataBytes[i], _, err = b.reader.Read()
+
+		if err != nil {
+			b.exhausted = true
+			// NB: errors other than EOF should halt execution.
+			if err != io.EOF {
+				b.err = err
+				return false
+			}
+
+			// NB: tag iters and data bytes are already released at this index;
+			// explicitly free other resources here.
+			b.recorders[i].release()
+			b.baseIters[i].Close()
+			b.byteReaders[i] = nil
+			b.err = b.freeAfterIndex(i + 1)
+			// NB: if any values remain, provide them to consumer.
+			return i > 0
+		}
+
+		b.dataBytes[i].IncRef()
+		b.bytesRefHeld[i] = true
+
+		bs := b.dataBytes[i].Bytes()
+		bbs := append(make([]byte, 0, len(bs)), bs...)
```

Comment:
The contents of `b.dataBytes[i].Bytes()` are copied here, and it looks like it is not being accessed anywhere else.
Any reason not to release the reference to it here immediately? Or perhaps it could even be made local within `Next()`?